### ISSUE TITLE
How to do real-time object detection？

### ISSUE THREAD

**Issue by Wuhanhan123** [ultralytics/yolov5#9726]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hi.
How to do real-time object detection？
Thanks :)

**Comment by glenn-jocher** [comment]
@Wuhanhan123 👋 Hello! Thanks for asking about **handling inference results**. YOLOv5 🚀 [PyTorch Hub](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading) models allow for simple model loading and inference in a pure python environment without using `detect.py`. 

### Simple Inference Example

This example loads a pretrained YOLOv5s model from PyTorch Hub as `model` and passes an image for inference. `'yolov5s'` is the YOLOv5 'small' model. For details on all available models please see the [README](https://github.com/ultralytics/yolov5#pretrained-checkpoints). Custom models can also be loaded, including custom trained PyTorch models and their [exported](https://docs.ultralytics.com/yolov5/tutorials/model_export) variants, i.e. ONNX, TensorRT, TensorFlow, OpenVINO YOLOv5 models.

```python
import torch

# Model
model = torch.hub.load('ultralytics/yolov5', 'yolov5s')  # yolov5n - yolov5x6 official model
#                                            'custom', 'path/to/best.pt')  # custom model

# Images
im = 'https://ultralytics.com/images/zidane.jpg'  # or file, Path, URL, PIL, OpenCV, numpy, list

# Inference
results = model(im)

# Results
results.print()  # or .show(), .save(), .crop(), .pandas(), etc.
results.xyxy[0]  # im predictions (tensor)

results.pandas().xyxy[0]  # im predictions (pandas)
#      xmin    ymin    xmax   ymax  confidence  class    name
# 0  749.50   43.50  1148.0  704.5    0.874023      0  person
# 2  114.75  195.75  1095.0  708.0    0.624512      0  person
# 3  986.00  304.00  1028.0  420.0    0.286865     27     tie

results.pandas().xyxy[0].value_counts('name')  # class counts (pandas)
# person    2
# tie       1
```
<img width="550" src="https://user-images.githubusercontent.com/26833433/149679662-b2021c78-59b2-4f52-b768-1e5161537c85.png">

See [YOLOv5 PyTorch Hub Tutorial](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading) for details.

Good luck 🍀 and let us know if you have any other questions!

**Comment by khelkun** [comment]
I raise up the question after several researches on the subject because I try to find how to achieve the fatest inference object detection with a Tensorflow.js model.

Those [benchmarks on the exported variants of YOLOv5](https://docs.ultralytics.com/yolov5/tutorials/model_export) are interesting, except there's none for Tensorflow.js (NaN). Would you have benchmarks about **Tensorflow.js**?

My experience in running Tensorflow.js inference in my web browser reveal this:
- For the **pre-trained YOLOv5s release 7.0** exported as tensorflow.js model (with export.py) I get around **60ms of inference time**.
- For a **custom trained YOLOv5s release 7.0** and exported as tensorflow.js model (with export.py) I get around **90ms of inference time**.
I assume the second one depends a lot of the training parameters I used but I basically followed the [roboflow collab notebook](https://github.com/roboflow/notebooks/blob/main/notebooks/train-yolov5-object-detection-on-custom-data.ipynb).

60ms is 16 FPS, 90ms is 11 FPS, how could I get closer to 30 FPS (e.g 33ms)?

**Comment by glenn-jocher** [comment]
@khelkun thanks for reaching out! Achieving faster inference with a Tensorflow.js model can be an iterative process involving both model architecture and the hardware it runs on. Here are a few optimizations to consider:

1. **Quantization**: Reducing model precision which lowers inference time at the cost of accuracy. 
2. **Model Pruning**: Removing unimportant weights to speed up inference while maintaining accuracy.
3. **Hardware Acceleration**: Using WebGL backend for Tensorflow.js which leverages the GPU for improved performance.
4. **Batch Inference**: Processing multiple inputs simultaneously can improve throughput.
5. **Model Architecture**: Experiment with lighter YOLOv5 variants or customizing model design based on specific requirements.

The community may provide further insights on optimizing Tensorflow.js inference speeds while maintaining model accuracy. All the best with your optimization efforts! 🚀
